src/link.ts

```diff
@@ -0,0 +1,234 @@
+import {
+  GraphQLRequest,
+  NextLink,
+  Operation,
+  RequestHandler,
+  FetchResult,
+} from './types';
+
+import {
+  validateOperation,
+} from './linkUtils';
+
+import {
+  parse,
+} from 'graphql/language/parser';
+
+import * as Observable from 'zen-observable';
+import { DocumentNode } from 'graphql/language/ast';
+
+export abstract class ApolloLink {
+
+  public static from(links: (ApolloLink | RequestHandler)[]) {
+    if (links.length === 0) {
+      return ApolloLink.empty();
+    }
+
+    return links
+      .map(ApolloLink.toLink)
+      .reduce((x, y) => x.concat(y));
+  }
+
+  public static empty(): ApolloLink {
+    return new FunctionLink((op, forward) => Observable.of());
+  }
+
+  public static passthrough(): ApolloLink {
+    return new FunctionLink((op, forward) => forward ? forward(op) : Observable.of());
+  }
+
+  // split allows for creating a split point in an execution chain
+  // like filter, it can be used to direct operations based
+  // on request information. Instead of dead ending an execution,
+  // split allows for new chains to be formed.
+  public static split(
+    test: (op: Operation) => boolean,
+    left: ApolloLink | RequestHandler,
+    right: ApolloLink | RequestHandler = ApolloLink.passthrough(),
+  ): ApolloLink {
+
+    left = ApolloLink.toLink(left);
+    right = ApolloLink.toLink(right);
+
+    if (ApolloLink.isTerminating(left) && ApolloLink.isTerminating(right)) {
+      return new TerminatedSplit(test, left, right);
+    }
+
+    return new SplitLink(test, left, right);
+  }
+
+  public static toLink(link: ApolloLink | RequestHandler): ApolloLink {
+    if (typeof link === 'function') {
+      return link.length <= 1 ? new TerminatedLink(link) : new FunctionLink(link);
```

Comment:
If we want to keep these here, we should use `.isTerminating`. But I don't think this helps since the function returns an `ApolloLink` anyway.